File: 操作系统/Windows API/Windows 程序设计/第6版/CPlusPlus/Chapter11/PetzoldProgrammingWindows6Chapter11/PetzoldProgrammingWindows6Chapter11/NewToggle.cpp
//
// NewToggle.cpp
// Implementation of the NewToggle class.
//

#include "pch.h"
#include "NewToggle.h"

using namespace PetzoldProgrammingWindows6Chapter11;

using namespace Platform;
using namespace Windows::Foundation;
using namespace Windows::Foundation::Collections;
using namespace Windows::UI::Xaml;
using namespace Windows::UI::Xaml::Controls;
using namespace Windows::UI::Xaml::Data;
using namespace Windows::UI::Xaml::Documents;
using namespace Windows::UI::Xaml::Input;
using namespace Windows::UI::Xaml::Interop;
using namespace Windows::UI::Xaml::Media;

DependencyProperty^ NewToggle::checkedContentProperty = 
    DependencyProperty::Register("CheckedContent",
                                 TypeName(Object::typeid),
                                 TypeName(NewToggle::typeid),
                                 ref new PropertyMetadata(nullptr));

DependencyProperty^ NewToggle::isCheckedProperty = 
    DependencyProperty::Register("IsChecked",
                                 TypeName(bool::typeid),
                                 TypeName(NewToggle::typeid),
                                 ref new PropertyMetadata(false,
                                    ref new PropertyChangedCallback(NewToggle::OnCheckedChangedStatic)));

NewToggle::NewToggle()
{
    DefaultStyleKey = "PetzoldProgrammingWindows6Chapter11.NewToggle";
}

void NewToggle::OnApplyTemplate()
{
    if (uncheckButton != nullptr)
        uncheckButton->Click -= uncheckClickEventToken;

    if (checkButton != nullptr)
        checkButton->Click -= checkClickEventToken; 

    uncheckButton = dynamic_cast<Button^>(GetTemplateChild("UncheckButton"));
    checkButton = dynamic_cast<Button^>(GetTemplateChild("CheckButton"));

    if (uncheckButton != nullptr)
        uncheckClickEventToken = uncheckButton->Click += ref new RoutedEventHandler(this, &NewToggle::OnButtonClick);

    if (checkButton != nullptr)
        checkClickEventToken = checkButton->Click += ref new RoutedEventHandler(this, &NewToggle::OnButtonClick);
}

void NewToggle::OnButtonClick(Object^ sender, RoutedEventArgs^ args)
{
    this->IsChecked = dynamic_cast<Button^>(sender) == checkButton;
}

void NewToggle::OnCheckedChangedStatic(DependencyObject^ obj, DependencyPropertyChangedEventArgs^ args)
{
    dynamic_cast<NewToggle^>(obj)->OnCheckedChanged(nullptr);
}

void NewToggle::OnCheckedChanged(Object^ args)
{
    VisualStateManager::GoToState(this, this->IsChecked ? "Checked" : "Unchecked", true);

    CheckedChanged(this, args);
}

File: 操作系统/Windows API/Windows 程序设计/第6版/CSharp/Chapter11/Petzold.ProgrammingWindows6.Chapter11/Petzold.ProgrammingWindows6.Chapter11/NewToggle.cs
using System;
using Windows.UI.Xaml;
using Windows.UI.Xaml.Controls;

namespace Petzold.ProgrammingWindows6.Chapter11
{
    public class NewToggle : ContentControl
    {
        public event EventHandler CheckedChanged;
        Button uncheckButton, checkButton;

        static NewToggle()
        {
            CheckedContentProperty = DependencyProperty.Register("CheckedContent",
                typeof(object),
                typeof(NewToggle),
                new PropertyMetadata(null));

            IsCheckedProperty = DependencyProperty.Register("IsChecked",
                typeof(bool),
                typeof(NewToggle),
                new PropertyMetadata(false, OnCheckedChanged));
        }

        public NewToggle()
        {
            this.DefaultStyleKey = typeof(NewToggle);
        }

        public static DependencyProperty CheckedContentProperty { private set; get; }

        public static DependencyProperty IsCheckedProperty { private set; get; }

        public object CheckedContent
        {
            set { SetValue(CheckedContentProperty, value); }
            get { return GetValue(CheckedContentProperty); }
        }

        public bool IsChecked
        {
            set { SetValue(IsCheckedProperty, value); }
            get { return (bool)GetValue(IsCheckedProperty); }
        }

        protected override void OnApplyTemplate()
        {
            if (uncheckButton != null)
                uncheckButton.Click -= OnButtonClick;

            if (checkButton != null)
                checkButton.Click -= OnButtonClick;

            uncheckButton = GetTemplateChild("UncheckButton") as Button;
            checkButton = GetTemplateChild("CheckButton") as Button;

            if (uncheckButton != null)
                uncheckButton.Click += OnButtonClick;

            if (checkButton != null)
                checkButton.Click += OnButtonClick;

            base.OnApplyTemplate();
        }

        void OnButtonClick(object sender, RoutedEventArgs args)
        {
 	        this.IsChecked = sender == checkButton;
        }

        static void OnCheckedChanged(DependencyObject obj,
                                     DependencyPropertyChangedEventArgs args)
        {
            (obj as NewToggle).OnCheckedChanged(EventArgs.Empty);
        }

        protected virtual void OnCheckedChanged(EventArgs args)
        {
            VisualStateManager.GoToState(this, 
                                         this.IsChecked ? "Checked" : "Unchecked", 
                                         true);

            if (CheckedChanged != null)
                CheckedChanged(this, args);
        }
    }
}
